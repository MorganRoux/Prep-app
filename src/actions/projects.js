import database from '../firebase/firebase';



export const createProject = (project) => ({
    type: 'CREATE_PROJECT',
    project
});

export const removeProject = (id) => ({
    type: 'REMOVE_PROJECT',
    id
});

export const setProjectData = (project) => ({
        type: 'SET_PROJECT',
        project
});

export const setCurrentProject = (project) => {
    return {
        type: 'SET_CURRENT_PROJECT',
        project
    }
}

export const noCurrentProject = () => {
    return {
        type: 'NO_CURRENT_PROJECT'
    }
}

export const startFetchProjectData = (projectId) => {
    return (dispatch) => {
        return database.ref(`/projects/${projectId}`).once('value')
        .then( (snapshot) => {
            
            const name = snapshot.val().name;
            const id = snapshot.key;
            // get staff data
            const staffData = [];
            snapshot.child('/staff').forEach( (childSnapshot) => {
                const { name, email, role } = childSnapshot.val();
                staffData.push({
                    id: childSnapshot.key,
                    name,
                    email,
                    role
                });
            });
            dispatch(setProjectData({
                id,
                name,
                staff: staffData
            }));

        });
    }
}

export const startRemoveProject = (projectId) => {
    return (dispatch, getState) => {
        const uid = getState().user.uid;

        return Promise.all([
            database.ref(`projects/${projectId}`).remove(),
            database.ref(`users/${uid}/projects/${projectId}`).remove()
        ]).then( () => dispatch(removeProject(projectId)));
    }
}

export const startCreateProject = () => {
    return (dispatch, getState) => {
        const {uid, profile} = getState().user;
        const {name, email} = profile
        const staff={};
        staff[uid] = {name, email, role:'5'}
        return database.ref('projects').push({
            name: 'New Project',
            staff
        })
        .then( (ref) => {

            return database.ref(`users/${uid}/projects/${ref.key}`)
            .set({
                name: 'New Project',
                role: '5'
            }).then( () => dispatch(createProject({
                name: 'New Project', 
                id: ref.key,
                staff: [{
                    uid,
                    name,
                    email,
                    role:'5'
                }]
            })));
        });
    }
    
}


