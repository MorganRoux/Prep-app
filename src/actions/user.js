import { firebase, googleAuthProvider } from '../firebase/firebase';
import database from '../firebase/firebase';
//import project from '../tests/fixtures/projects';

export const startLogin = () => {
    //no need to be here, should stay with loginpage
    return () => {
        return firebase.auth().signInWithPopup(googleAuthProvider);
    }
}

export const startLogout = () => {
    //no need to be here, should stay with equipemntheader
    return () => {
        return firebase.auth().signOut();
    }
}

export const login = (uid) => {
    return {
        type: 'LOGIN',
        uid
    }

}

export const logout = (uid) => {
    return {
        type: 'LOGOUT'
    }
}

export const setUserData = (user) => {
    return {
        type: 'SET_USER_DATA',
        user
    }
}

export const setEquipementRef = (ref) => {
    return {
        type: 'SET_REF',
        ref
    }
}

export const startFetchUserData = (uid) => {
    return (dispatch) => {
        return database.ref(`users/${uid}`)
        .once('value')
        .then( (snapshot) => {

            //test if the user has already logged in
            if( !snapshot.exists() ) {
                throw Error('User not found');
            }

            const {currentProject, profile} = snapshot.val();
            const projectsData = [];
            snapshot.child('/projects').forEach((childSnapshot) => {

                projectsData.push({
                    id: childSnapshot.key,
                    name: childSnapshot.val().name,
                    role: childSnapshot.val().role
                });
            });
            dispatch(setUserData({
                uid,
                currentProject,
                profile,
                projects : projectsData
            }));
        });
    }
}

export const createProfile = (uid) => {
    return (dispatch) => {
        const newProfile = {
            uid,
            currentProject: 'idp1',
            profile: {
                name: 'Anonymous',
                email: 'unknown'
            },
            projects: {
                'idp1' : {
                    name: 'projet1',
                    role: '5'
                }
            }
        };
        return database.ref(`users/${uid}`).set(newProfile)
        .then( () => {
            dispatch(setUserData({
                uid,
                currentProject: 'idp1',
                profile: {
                    name: 'Anonymous',
                    email: 'unknown'
                },
                projects: [{
                    id: 'idp1',
                    name: 'projet1',
                    role: '5'
                }]
            })
        )});
    }
}