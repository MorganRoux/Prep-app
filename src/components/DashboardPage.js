import React from 'react';
import EquipmentList from './EquipmentList';

const DashboardPage = () => {
    return (
        <div>
            <h1>Dashboard Page</h1>
            <EquipmentList />
        </div>
        
    );
};

export default DashboardPage;