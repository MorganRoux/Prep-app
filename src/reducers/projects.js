

const projectsDefaultState = null; //projects;

const projectReducer = ( state = projectsDefaultState, action ) => {

    switch(action.type) {
    
    case 'REMOVE_PROJECT' :
        return null;

    case 'SET_PROJECT' : 
        return action.project;

    case 'CREATE_PROJECT' :
        return action.project;
        
    case 'SET_CURRENT_PROJECT': 
        const {id, name, staff: staff0} = action.project;
        const staff = Object.keys(staff0).map( (key) => ({id: key, ...staff0[key]}));
        return {id, name, staff}

    case 'NO_CURRENT_PROJECT' :
        return null

    default:
        return state;
    }
}

export default projectReducer