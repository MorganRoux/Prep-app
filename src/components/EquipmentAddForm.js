import React from 'react'
import { Paper, TableBody, TableRow, TableCell, TableFooter } from '@material-ui/core';
import IconButton from '@material-ui/core/IconButton';
import Icon from '@material-ui/core/Icon';

import { startAddEquipment } from '../actions/equipment'

import { connect } from 'react-redux'
import { withSnackbar } from 'notistack'

export class EquipmentAddForm extends React.Component {
    
    constructor(props) {
        super(props);

        this.state = {
            quantity: '',
            publicName: ''
        }
    }

    handleCloseSnackBar = (event, reason) => {
        if (reason === 'clickaway') {
          return;
        }
    
        this.setState({ snackBarOpen: false });
      };

    onQuantityChange = (e) => {
        const quantity = e.target.value;
        if(quantity.match(/^\d*$/)) {
            this.setState( () => ({quantity}));
        }
    }

    onPublicNameChange = (e) => {
        const publicName = e.target.value;
        this.setState( () => ({publicName}));
    }

    onAdd = (e) => {
        e.preventDefault()
        // test the inputs
        if(!this.state.publicName || !this.state.quantity) {
            this.props.enqueueSnackbar('Informations erronnées', {variant:'error'});
            return;
        }

        // search the item in the stocklist
        const itemToAdd = this.props.stocklist.find( (item) => {
            return (item.publicName === this.state.publicName)
        });
        if (!itemToAdd) { 
            this.props.enqueueSnackbar('Élement introuvable', {variant:'error'});
            return;
        }

        // add item
        this.props.startAddEquipment({
            ...itemToAdd,
            quantity: this.state.quantity
        });
        const s = this.state.quantity === '1' ? '':'s'
        this.props.enqueueSnackbar(
            `${this.state.quantity}x ${this.state.publicName} ajouté${s}`, {variant:'success'});

        this.setState( () => ({
            quantity: '',
            publicName:''
        }));
    }

    render () {
        return (
            <TableFooter>
                <TableRow className = "addForm-row">
                    <TableCell />
                    <TableCell colSpan="4">
                        <form onSubmit={this.onAdd}>
                        <input 
                        type="text"
                        name= "quantity"
                        autoFocus
                        placeholder=""
                        value={this.state.quantity}
                        onChange = {this.onQuantityChange}
                        />
                 
                    
                        <input 
                        type="text"
                        name= "publicName"
                        placeholder=""
                        value={this.state.publicName}
                        onChange = {this.onPublicNameChange}
                        />
                    
                        <IconButton className="add-button" aria-label="Add" type="submit">
                            <Icon color="primary">
                            add_circle
                            </Icon>
                        </IconButton>
                        </form>
                    </TableCell>
                </TableRow>
                </TableFooter>
        );
    }
}

const mapStateToProps = (state) => {
    return {
        equipements: state.equipements,
        stocklist: state.stocklist,
    }
}

const mapDispatchToProps = (dispatch) => {
    return {
        startAddEquipment: (item) => dispatch(startAddEquipment(item))
    }
}

export default connect(mapStateToProps, mapDispatchToProps)(withSnackbar(EquipmentAddForm))