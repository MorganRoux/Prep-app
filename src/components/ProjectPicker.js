import React from 'react';
import { connect } from 'react-redux';
import Select from '@material-ui/core/Select';
import { startRemoveProject, startCreateProject, setCurrentProject, noCurrentProject } from '../actions/projects';
import database from '../firebase/firebase';

export class ProjectPicker extends React.Component { 

    setCurrentProject = (id) => {
            return database.ref(`/projects/${id}`).once('value')
            .then( (snapshot)=> {
                const projectToSet = snapshot.val();
                this.props.setCurrentProject({...projectToSet, id});
            });
    }

    onChange = (e) => {
        switch(e.target.value) {
        case 'select' : 
            return;
        case 'noproject' : 
            return;
        case 'remove' :    
            this.props.startRemoveProject(this.props.user.currentProject);
            this.props.noCurrentProject();
            return;

        case 'create' :
            this.props.startCreateProject();
            return;

        default:
            this.setCurrentProject(e.target.value)
            return;
        }
    }

    render() {    
        return (
                    <Select
                        native
                        onChange={this.onChange}
                        value={this.props.user.currentProject}
                    >
                    {this.props.user.projects.length > 0 ? (
                        <option key="option-title-select" value="select">Selectionner</option>
                         
                    ) : (
                        <option key="option-title-noproject" value="noproject">Aucun Projet</option>
                    )}
                    
                    { this.props.user.projects.map( (project) => (
                        <option 
                            key = {project.id} 
                            value={project.id}
                        >
                        {project.name}
                        </option>
                    ))}
                    <option key="option-create" value="create">Cr√©er</option>
                    {this.props.user.projects.length > 0 && (
                        <option key="option-remove" value="remove">Supprimer</option>
                    )}
                    </Select>
        );
    }
    
};

const mapStateToProps = (state) => ({
    user: state.user,
    project: state.project
});

const mapDispatchToProps = (dispatch) => ({
    setCurrentProject: (id) => dispatch(setCurrentProject(id)),
    startRemoveProject: (id) => dispatch(startRemoveProject(id)),
    startCreateProject: (id) => dispatch(startCreateProject()),
    noCurrentProject: () => dispatch(noCurrentProject())
})
export default connect(mapStateToProps,mapDispatchToProps)(ProjectPicker);