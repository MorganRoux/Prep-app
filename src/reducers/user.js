import uuid from 'uuid'
import { addKitActions } from '../tests/fixtures/actions';
const userDefaultState = {}
let projects =[];

const userReducer = (state = userDefaultState, action) => {
    switch(action.type) {
    case 'LOGIN' : 
        return { ...state, uid: action.uid};
    
    case 'LOGOUT' :
        return {...state, uid :''};

    case 'SET_CURRENT_PROJECT' :
        return {...state, currentProject: action.id};

    case 'SET_USER_DATA':
        return action.user;

    case 'REMOVE_PROJECT': 
        projects = state.projects.filter( ({id}) => action.id !== id);
        return {...state, projects};

    case 'CREATE_PROJECT' :
        const {name, id } = action.project;
        
        if(state.projects !== undefined) {
        projects = [...state.projects, {name, id, role:'5'}]
        } else {
        projects = [{name, id, role:'5'}];
        }
         return {...state, projects}

    default:
        return state
    }

}

export default userReducer;