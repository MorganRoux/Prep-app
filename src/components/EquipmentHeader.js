import React from 'react'
import { connect } from 'react-redux';
import InputBase from '@material-ui/core/InputBase';
import IconButton from '@material-ui/core/IconButton';
import MenuIcon from '@material-ui/icons/Menu';
import SearchIcon from '@material-ui/icons/Search';
import { setTextFilter } from '../actions/filters';
import { startLogout } from '../actions/user';

export class EquipmentHeader extends React.Component {

    constructor (props) {
        super(props);
        this.state = {
            searchText : ''
        }
    }

    onSearchChange = (e) => {

        const searchText = e.target.value;
        this.setState(() => ({searchText}));
        this.props.setTextFilter(searchText);
    }
    
    render () {
        return (
            <div style = {{display: 'flex', justifyContent: 'space-between'}}>
                <div style = {{display: 'flex', justifyContent: 'flex-start'}}>
                    <IconButton>
                        <MenuIcon />
                    </IconButton>
                    <h1>Equipment List</h1>
                </div>
                <div style = {{display: 'flex', justifyContent: 'flex-end'}}>
                    <InputBase 
                        className="search-field"
                        placeholder="Rechercher" 
                        value={this.state.searchText}
                        onChange={this.onSearchChange}
                    />
                    <IconButton>
                        <SearchIcon />
                    </IconButton>
                    <button className = "logout-button" onClick={this.props.startLogout}>Logout</button>
                </div>
            </div>
    )};

}

const mapDispatchToProps = (dispatch) => ({
    setTextFilter : (text) =>  dispatch(setTextFilter(text)),
    startLogout : () => dispatch(startLogout())
});

export default connect(null, mapDispatchToProps)(EquipmentHeader)