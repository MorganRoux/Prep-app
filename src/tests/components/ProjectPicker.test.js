import React from 'react';
import { ProjectPicker} from '../../components/ProjectPicker';
import user from '../fixtures/user';
import { project } from '../fixtures/projects'
import { shallow } from 'enzyme';
import Select from '@material-ui/core/Select';
import { setupFirebase } from '../fixtures/firebase';

let setCurrentProject, startRemoveProject, startCreateProject, noCurrentProject, wrapper;

beforeEach( (done) => {
    setCurrentProject = jest.fn();
    startRemoveProject = jest.fn();
    startCreateProject = jest.fn();
    noCurrentProject = jest.fn();
    wrapper = shallow(
        <ProjectPicker 
            setCurrentProject={setCurrentProject}
            startRemoveProject={startRemoveProject}
            startCreateProject={startCreateProject}
            noCurrentProject={noCurrentProject}
            user={user}
            project={project}
        />);

    setupFirebase().then( () => {
        done(); 
    }); 
});

test('should render ProjectPicker', () => {
    expect(wrapper).toMatchSnapshot();
});

test('should handle select change', (done) => {
    const id = 'idp1'
    wrapper.find(Select).simulate('change', {target:{value: id}})
        expect(startRemoveProject).not.toHaveBeenCalled();
        expect(startCreateProject).not.toHaveBeenCalled();
        done();
    });

test(' should retrieve the project from the database', (done) => {
    const id = 'idp3';
    wrapper.instance().setCurrentProject(id).then( () => {
        expect(setCurrentProject).toHaveBeenCalledWith({
            id: 'idp3',
            name: 'projet3',
            staff: {
                '1625HT28': {
                    name: 'Morgan',
                    email: 'mail@mail.com',
                    role: '5'
                },
                'oiiu567' : {
                    name: 'Morgan2',
                    email: 'mail2@mail.com',
                    role: '5'
                }
            }
        });
        done();
    });
});

test('should handle remove selection', () => {
    wrapper.find(Select).simulate('change', {target:{value: 'remove'}});
    expect(noCurrentProject).toHaveBeenCalled();
    expect(startRemoveProject).toHaveBeenCalled();
    expect(startCreateProject).not.toHaveBeenCalled();

});

test('should handle create selection', () => {
    wrapper.find(Select).simulate('change', {target:{value: 'create'}});
    expect(setCurrentProject).not.toHaveBeenCalled()
    expect(startRemoveProject).not.toHaveBeenCalled();
    expect(startCreateProject).toHaveBeenCalled();
})

