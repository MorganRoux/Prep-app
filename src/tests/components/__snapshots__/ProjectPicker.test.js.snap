// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ProjectPicker 1`] = `
<WithStyles(WithFormControlContext(Select))
  native={true}
  onChange={[Function]}
  value="idp1"
>
  <option
    key="option-title-select"
    value=""
  >
    Selectionner
  </option>
  <option
    key="idp1"
    value="idp1"
  >
    projet1
  </option>
  <option
    key="idp2"
    value="idp2"
  >
    projet2
  </option>
  <option
    key="option-create"
    value="create"
  >
    Cr√©er
  </option>
  <option
    key="option-remove"
    value="remove"
  >
    Supprimer
  </option>
</WithStyles(WithFormControlContext(Select))>
`;
